server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java

```diff
@@ -1081,18 +1156,49 @@ protected void writeIndexGen(RepositoryData repositoryData, long expectedGen, bo
                 genBytes = bStream.bytes();
             }
             logger.debug("Repository [{}] updating index.latest with generation [{}]", metadata.name(), newGen);
+
+            // Step 3: Update CS to reflect new repository generation.
             writeAtomic(INDEX_LATEST_BLOB, genBytes, false);
-            // delete the N-2 index file if it exists, keep the previous one around as a backup
-            if (newGen - 2 >= 0) {
-                final String oldSnapshotIndexFile = INDEX_FILE_PREFIX + Long.toString(newGen - 2);
-                try {
-                    blobContainer().deleteBlobIgnoringIfNotExists(oldSnapshotIndexFile);
-                } catch (IOException e) {
-                    logger.warn("Failed to clean up old index blob [{}]", oldSnapshotIndexFile);
+            clusterService.submitStateUpdateTask("set safe repository generation", new ClusterStateUpdateTask() {
+                @Override
+                public ClusterState execute(ClusterState currentState) {
+                    final RepositoriesState state = currentState.custom(RepositoriesState.TYPE);
+                    final RepositoriesState.State repoState = state.state(metadata.name());
+                    final long prevGeneration = repoState.generation();
+                    if (prevGeneration != expectedGen) {
+                        throw new IllegalStateException("Tried to update repo generation to [" + newGen
+                            + "] but saw unexpected generation in state [" + repoState + "]");
+                    }
+                    if (repoState.pendingGeneration() == prevGeneration) {
+                        throw new IllegalStateException(
+                            "Tried to update non-pending repo state [" + repoState + "] after write to generation [" + newGen + "]");
+                    }
+                    final RepositoriesState updated =
+                        RepositoriesState.builder().putAll(state).putState(metadata.name(), newGen, newGen).build();
+                    return ClusterState.builder(currentState).putCustom(RepositoriesState.TYPE, updated).build();
                 }
-            }
-            return null;
-        });
+
+                @Override
+                public void onFailure(String source, Exception e) {
+                    l.onFailure(e);
```

Comment:
should we wrap the exception here to add more context where exactly something went wrong (same for setting pending generation)?